// unframer/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResolveLinks = exports.Link = exports.UnframerProvider = exports.AdaptedLink = exports.ContextProviders = exports.WithFramerBreakpoints = exports.FramerStyles = void 0;
__exportStar(require("./framer.js"), exports);
const react_js_1 = require("./react.js");
Object.defineProperty(exports, "Link", { enumerable: true, get: function () { return react_js_1.AdaptedLink; } });
Object.defineProperty(exports, "ResolveLinks", { enumerable: true, get: function () { return react_js_1.ResolveLinksAdapted; } });
var react_js_2 = require("./react.js");
Object.defineProperty(exports, "FramerStyles", { enumerable: true, get: function () { return react_js_2.FramerStyles; } });
Object.defineProperty(exports, "WithFramerBreakpoints", { enumerable: true, get: function () { return react_js_2.WithFramerBreakpoints; } });
Object.defineProperty(exports, "ContextProviders", { enumerable: true, get: function () { return react_js_2.ContextProviders; } });
Object.defineProperty(exports, "AdaptedLink", { enumerable: true, get: function () { return react_js_2.AdaptedLink; } });
Object.defineProperty(exports, "UnframerProvider", { enumerable: true, get: function () { return react_js_2.UnframerProvider; } });
