// unframer/dist/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.dedent = exports.stackblitzDemoExample = exports.spinner = exports.logger = void 0;
exports.terminalMarkdown = terminalMarkdown;
exports.componentNameToPath = componentNameToPath;
exports.sleep = sleep;
exports.kebabCase = kebabCase;
exports.isTruthy = isTruthy;
const string_dedent_1 = __importDefault(require("string-dedent"));
exports.dedent = string_dedent_1.default;
const picocolors_1 = __importDefault(require("picocolors"));
const nanospinner_1 = require("nanospinner");
function terminalMarkdown(markdown) {
    return markdown;
}
const shouldDebugUnframer = !!process.env.DEBUG_UNFRAMER;
const prefix = '[unframer]';
exports.logger = {
    debug: shouldDebugUnframer,
    log(...args) {
        if (!exports.logger.debug) {
            return;
        }
        console.log(prefix, ...args);
    },
    start(x) {
        if (!x)
            return;
        console.log(prefix, x);
    },
    stop(x) {
        if (!x)
            return;
        console.log(prefix, x);
    },
    info(...args) {
        console.log(prefix, ...args);
    },
    update(...args) {
        console.log(prefix, ...args);
    },
    green(...args) {
        console.log([prefix, ...args].map((x) => picocolors_1.default.green(x)).join(' '));
    },
    error(...args) {
        console.error([prefix, ...args].map((x) => picocolors_1.default.red(x)).join(' '));
    },
};
exports.spinner = (0, nanospinner_1.createSpinner)('Unframer');
function componentNameToPath(name) {
    return (name
        .split('/')
        .filter(Boolean)
        .map((part) => kebabCase(part))
        .join('/') || 'component-without-name');
}
function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
// https://www.npmjs.com/package/just-kebab-case?activeTab=readme
// any combination of spaces and punctuation characters
// thanks to http://stackoverflow.com/a/25575009
var wordSeparators = /[\s\u2000-\u206F\u2E00-\u2E7F\\'!"#$%&()*+,\-.\/:;<=>?@\[\]^_`{|}~]+/;
var capital_plus_lower = /[A-ZÀ-Ý\u00C0-\u00D6\u00D9-\u00DD][a-zà-ÿ]/g;
var capitals = /[A-ZÀ-Ý\u00C0-\u00D6\u00D9-\u00DD]+/g;
function kebabCase(str) {
    // replace word starts with space + lower case equivalent for later parsing
    // 1) treat cap + lower as start of new word
    str = str.replace(capital_plus_lower, function (match) {
        // match is one caps followed by one non-cap
        return ' ' + (match[0].toLowerCase() || match[0]) + match[1];
    });
    // 2) treat all remaining capitals as words
    str = str.replace(capitals, function (match) {
        // match is a series of caps
        return ' ' + match.toLowerCase();
    });
    return str
        .trim()
        .split(wordSeparators)
        .join('-')
        .replace(/^-/, '')
        .replace(/-\s*$/, '');
}
function isTruthy(x) {
    return Boolean(x);
}
exports.stackblitzDemoExample = process.env.STACKBLITZ_DEMO_EXAMPLE;
